concepts/type-theory-glossary.md

```diff
@@ -0,0 +1,396 @@
+
+Consider three boolean values: how many states can it have?
+(You can draw the full table yourself, for this one.)
+When you count them all up, you should reach `8`.
+
+But now, let's imagine something new: suppose we've got some sort of values
+that we call "nooleans".  Nooleans can have the value 'nue' or 'fuu'.
+
+We'll explore more examples of this, and start identifying categories of
+things where we can do predictable kinds of math with cardinality in
+the next section: [Categories of Types](#categories-of-types).
+(Spoiler: the examples above correspond to an enum, then a product type,
+then another bigger product type, and finally a sum type.)
+
+
+Categories of Types
+-------------------
+
+We can identify a few categories of "types" based on how we evaluate their cardinality.
+(These definitions will use names that are popularized in computer science,
+and so they *may* match the name used in your favorite programming language -- but
+whether or not the same is familiar, the *concept* is universal, even across languages!)
+
+### enums
+
+Enums have a cardinality that is some integer.
+
+For example, an enum might have members named "Foo", "Bar", and "Baz";
+and the cardinality of that enum is "|3|".
+
+(The term "enum" might make more sense if you think of it as short for
+*enumerated*, rather than *enumerable* -- something with countable-infinity
+members is **not** considered an "enum".
+(We'd call that an [infinite scalar](#infinite-scalar) instead.))
+
+### sum types
+
+Sum types have a cardinality that is *the sum of cardinality of their member types*.
+
+Sum types usually show up in prose descriptions of data as the word "or".
+
+For example, say: "I have a boolean *or* an int" --
+this would be an example of a sum type,
+and its cardinality would be "|inf+2|" --
+countable infinity for the int, and plus two for the boolean.
+
+A clearer example might take two enums -- let's call them
+"Animals" (it has members "cat", "dog", "rabbit"),
+and "Furniture" (it has members "chair", and "desk") --
+and say "I have an Animal *or* a furniture" --
+then clearly I have one (and only one) of these five things:
+the cardinality is "|3+2|".
+
+Sum types go by many names in various communities: they're also commonly
+known as "(discriminated) unions", "(tagged) variants", or other terminology.
+(We use the "union" term ourselves, in the lexicon of IPLD Schemas!)
+
+You can also consult the Wikipedia page on
+[Sum Types](https://en.wikipedia.org/wiki/Sum_types)
+for broader discussion of the topic.
+https://chadaustin.me/2015/07/sum-types/ also contains some useful content
+and examples from a variety of languages.
+
+### product types
+
+Product types have a cardinality that is *the multiplied cardinality of their member types*.
+
+This may sound outlandish, but you've been using them all your life:
+*"product types" are commonly known as "structs"*
+(or, "classes", in more aggressively object-oriented languages).
```

Comment:
"Classes" are often more likely to be infinite-recursive than structs. Heck, in C#, "capable of being self-recursive" would be a reasonable way to distinguish `struct`s from `class`es!

So I'd maybe just state that classes _can_ be product types, rather than implying that the definition of "class" is equivalent to "product type".